Extension/src/LanguageServer/settings.ts

```diff
@@ -311,90 +312,205 @@ export class CppSettings extends Settings {
+                if (setting.items.enum !== undefined) {
+                    if (!value.every(x => this.validateEnum(setting.items.enum, x))) {
+                        return setting.default;
+                    }
+                }
+            }
+            return value;
+        }
+        return setting.default;
+    }
+
+    private getAsExcludes(settingName: string): Excludes {
+        const value: any = super.Section.get(settingName);
+        if (isValidMapping(value, 'string', 'boolean')) {
+            return value as Excludes;
+        }
+        const setting = getRawSetting("C_Cpp." + settingName);
+        return setting.default as Excludes;
+    }
+
+    private getAsAssociations(settingName: string): Associations {
+        const value: any = super.Section.get(settingName);
+        if (isValidMapping(value, 'string', 'string')) {
+            return value as Associations;
+        }
+        const setting = getRawSetting("C_Cpp." + settingName);
+        return setting.default as Associations;
+    }
+
+    // Checks a given enum value against a list of valid enum values from package.json.
+    private validateEnum(enumDescription: any, value: any): boolean {
+        if (isArray(enumDescription) && enumDescription.length > 0) {
+            return enumDescription.some(x => x.toLowerCase() === value.toLowerCase());
+        }
+        return false;
+    }
+
+    // Validates whether the given value is a valid enum value for the given setting.
+    private isValidEnum(settingName: string, value: any): boolean {
+        const setting = getRawSetting("C_Cpp." + settingName);
+        if (this.validateEnum(setting.enum, value)) {
+            return value;
+        }
+        return setting.default;
+    }
+
+    private isArrayOfCommentContinuationPatterns(x: any): x is (string | CommentPattern)[] {
+        return isArray(x) && x.every(y => isString(y) || this.isCommentPattern(y));
+    }
+
+    private isCommentPattern(x: any): x is CommentPattern {
+        return isString(x.begin) && isString(x.continue);
+    }
+
+    public get maxConcurrentThreads(): number | null { return this.getAsNumber("maxConcurrentThreads", true); }
+    public get maxMemory(): number | null { return this.getAsNumber("maxMemory", true); }
+    public get maxSymbolSearchResults(): number { return this.getAsNumber("maxSymbolSearchResults"); }
+    public get maxCachedProcesses(): number { return this.getAsNumber("maxCachedProcesses"); }
+    public get intelliSenseMaxCachedProcesses(): number | null { return this.getAsNumber("intelliSense.maxCachedProcesses", true); }
+    public get intelliSenseMaxMemory(): number | null { return this.getAsNumber("intelliSense.maxMemory", true); }
+    public get referencesMaxConcurrentThreads(): number | null { return this.getAsNumber("references.maxConcurrentThreads", true); }
+    public get referencesMaxCachedProcesses(): number { return this.getAsNumber("references.maxCachedProcesses"); }
```

Comment:
same here

Reply:
We discussed earlier that only settings that default to null should return null. Otherwise, we shouldn't accept a null value (https://github.com/microsoft/vscode-cpptools/pull/12454#discussion_r1676295154) It may be more correct to remove null as a possible value for this setting. 

@bobbrow what do you think?

Reply:
It looks like this setting allows `null` according to package.json. It doesn't matter if it's the default value. It's a possible value so this should use the overload to get `number | null`.

Reply:
Should we make null the default value for the setting as well?

Reply:
We are not changing `package.json` in this PR. We can discuss it as a follow up.